components/resource_metering/src/cpu/reporter/endpoint/grpc.rs

```diff
@@ -0,0 +1,109 @@
+use tikv_util::warn;
+
+pub struct GRPCEndpoint {
+    env: Arc<Environment>,
+    address: String,
+    client: Option<ResourceUsageAgentClient>,
+    limiter: Limiter,
+}
+
+impl Endpoint for GRPCEndpoint {
+    fn report(&mut self, address: &str, records: Records) {
+        if address.is_empty() {
+            return;
+        }
+
+        let handle = self.limiter.try_acquire();
+        if handle.is_none() {
+            return;
+        }
+
+        if self.address != address || self.client.is_none() {
+            self.address = address.to_owned();
+            self.init_client();
+        }
+
+        let client = self.client.as_ref().unwrap();
+        let call_opt = CallOption::default().timeout(Duration::from_secs(2));
+        let call = client.report_cpu_time_opt(call_opt);
+        if let Err(err) = &call {
+            warn!("failed to connect to agent"; "error" => ?err);
+            return;
+        }
+
+        let (mut tx, rx) = call.unwrap();
+        client.spawn(async move {
+            let _hd = handle;
+
+            let others = records.others;
+            let records = records.records;
+            for (tag, (timestamp_list, cpu_time_ms_list, _)) in records {
+                let mut req = CpuTimeRecord::default();
+                req.set_resource_group_tag(tag);
+                req.set_record_list_timestamp_sec(timestamp_list);
+                req.set_record_list_cpu_time_ms(cpu_time_ms_list);
+                if let Err(err) = tx.send((req, WriteFlags::default())).await {
+                    warn!("failed to send cpu records"; "error" => ?err);
+                    return;
+                }
+            }
+
+            // others
+            if !others.is_empty() {
+                let timestamp_list = others.keys().cloned().collect::<Vec<_>>();
+                let cpu_time_ms_list = others.values().cloned().collect::<Vec<_>>();
+                let mut req = CpuTimeRecord::default();
+                req.set_record_list_timestamp_sec(timestamp_list);
+                req.set_record_list_cpu_time_ms(cpu_time_ms_list);
+                if let Err(err) = tx.send((req, WriteFlags::default())).await {
+                    warn!("failed to send cpu records"; "error" => ?err);
+                    return;
+                }
+            }
+
+            if let Err(err) = tx.close().await {
+                warn!("failed to close a grpc call"; "error" => ?err);
+                return;
+            }
+
+            if let Err(err) = rx.await {
+                warn!("failed to receive from a grpc call"; "error" => ?err);
+            }
+        });
+    }
+}
+
+impl GRPCEndpoint {
+    fn init_client(&mut self) {
+        let channel = {
+            let cb = ChannelBuilder::new(self.env.clone())
+                .keepalive_time(Duration::from_secs(10))
+                .keepalive_timeout(Duration::from_secs(3));
+            cb.connect(&self.address)
```

Comment:
Can this tcp connection be reconnected when the server(agent) down or restarts? If not, what I understand is that all gRPC requests will fail after a restart from the server side, because the connection has been broken. So do we need to add a complex connection management mechanism for this?

Reply:
I think it will reconnect. The [integration test](https://github.com/tikv/tikv/blob/5c887790bfbe2487f479a5fa964577f415f247e2/tests/integrations/resource_metering/test_agent.rs#L136-L199) simulated the situation you described.